### DISCUSSION TITLE
Combined Optimizers & Comms Overlap

### DISCUSSION THREAD

**Discussion by chrisjmccormick** [KellerJordan/modded-nanogpt#170] [Show and tell]
I thought I'd make a thread to serve as a kind of blog post for this topic. Here's what I've figured out, and what I have implemented currently.

Here's the current overall picture.

<img width="2120" height="342" alt="image" src="https://github.com/user-attachments/assets/a26f38f2-52e7-4962-820c-3bf91340bcb9" />

To start, here's what I understand of the problems. 

## Issues

Note that, while it's nice to see the compute happening as quickly as possible, because we're communication bound I think there may only be two issues to solve in order to get the most overlap possible.

The first is memcpys related to our batching strategy.

### Batched vs. Sharded

Muon currently uses a "batched" strategy, where different matrices (e.g., from different layers) are stacked together, then distributed across the GPUs, so each GPU is working on a different set of matrices. 

The batches are determined by matrix shape. You can stack together whatever matrices, so long as they're the same shape. We currently leverage this, for example, to execute a batch that has both attention and mlp weights in it.

The "same shape" requirement actually comes from the communication backend, not from Muon itself. 

**Stacking and Unstacking**

The downside to the batched approach is that stacking step required before sending the data, and then unstacking again at the end.

It took me a while to understand what these green hash marks are. I knew they were memcpys, but I wasn't sure what for. I just wasn't familiar enough with the code yet. They represent the stacking and unstacking.

<img width="2119" height="407" alt="image" src="https://github.com/user-attachments/assets/a6aa8ec2-2a38-4be6-974d-eb9bb99b08df" />

The biggest delay they cause is at the end, where we have to unstack the MLP weights before we can start the next forward pass. 

I think they may also be causing a delay in sending out the attention weights (the "Comms Delay" at the beginning), because it's taking too long to stack? This results in the compute delay in the middle, where we're starting work on the attention update later than we could be. (Though you could also attribute that to the smear gate or attention gates).

### Gradient Accumulation

The other problem area is this region that immediately follows the backward pass.

<img width="1962" height="548" alt="image" src="https://github.com/user-attachments/assets/56395e12-a8dd-4b1a-a1c7-2e0c9dc12728" />

I saw @shenberg point this out in a recent comment as well--this is the gradient accumulation step.

It's an extra step incurred by running backward multiple times on a set of weights before consuming the gradients. 

The first time you run backward, the gradients are stored directly into the `.grad` property and there is no extra step required. If you run backward again, it requires extra compute to sum the new grads with the old ones.

When there are less than 8 GPUs, we use 'grad_accum' steps to reduce the batch size. The reason we see the above even in the 8x GPU setup, though, is because we only consume the Adam gradients every other step, so we have to do this  extra accumulation work.

This wouldn't necessarily be a problem since we have plenty of communication to do to overlap with this, except for the challenge of getting those scatters in flight.

### Backwards Hooks

To get the Adam gradients moving sooner, @akash5474 cleverly leveraged the `torch.Tensor.register_post_accumulate_grad_hook` feature to issue the scatter command on a tensor as soon as its ready.

What's been driving me nuts is that I can't get it to scatter any of the Muon weights this way. Because we don't accumulate gradients for the Muon weights, they're actually the ones that we should be able to send immediately!

I finally figured out the reason for this. It's because the `_post_accumulate_grad_hook` is literally only called _after a gradient accumulation_ operation. 

You'd expect it to mean "as soon as the gradients are ready", but (at least in our codebase?) that's not the case.

I've confirmed this in part by running on a 2x GPU setup where the MLP weights do have to be accumulated, and sure enough, the hook fired. 

I can try profiling without `torch.compile` to see if that's somehow related. 

The PyTorch docs [here](https://docs.pytorch.org/docs/stable/notes/autograd.html#whether-a-particular-hook-will-be-fired) don't speak to the issue, and there doesn't appear to be a different hook we can use.

### LM Head

Fortunately, we're getting pretty good overlap of the gradient accumulation  already with Akash's improvement. 

There's one spot where this is particularly frustrating, though.

First off, note that the Value Embeddings, Input Embeddings, and LM Head are all the same shape and data type. And yet, look how long it takes to accumulate its gradients--it's 350us compared to 80us for the other parameters.

The most annoying part of this is that this kernel runs _every step_.

<img width="2081" height="517" alt="image" src="https://github.com/user-attachments/assets/5c4df5c8-ac4c-4632-8312-2aa1ad68aa97" />

In a Muon-only step, no one needs to run accumulation, because the first backwards pass is always "free". I'm unclear on whether we're stuck with this or not--it could be that the memory layout problems force it, or that the custom backwards implementation isn't doing what it needs to to avoid this. 

Gemini seems to get it, I can drop its assessment in a separate comment further down. I'll try out its ideas.

## Solutions

Those are the problems; here's what I've come up with to solve the first few.

### Parameter Bank

Notice how Adam does not require this parameter stacking step. It's because it's able to use `all_gather_into_tensor`--an all gather operation where the backend is able to write the data from the other GPUs directly into the tensor, without a separate memcpy step. 

The reason its able to do that is Adam works on a per-element level, which means we can shard the tensors and have each GPU working on a different slice. 

Muon requires the entire parameter on a single GPU in order to orthogonalize it.

In order to use the slicing strategy, we can change how we store our weights--we stack them all together into "parameter banks".

Instead of each MLP having its own weights, the GPT module holds a (11, 2, 3072, 768) matrix, and each MLP is passed a view into that matrix. 

To slice this properly, we have to tack on a dummy mlp and make that (12, 2, 3072, 768) so that it's 24 matrices, and each GPU gets 3 mlp matrices to work on.

Attention already fits nicely. Its bank is (10, 3072, 768), but because you can break it into QKVO, it can be viewed as (40, 768, 768), and each GPU gets 5 matrices to work on.

The benefits are:

1. We avoid all of the stacking and unstacking, and the GPUs are only ever working on just attention weights or just mlp weights, so there's less imbalance in workload.
2. It decouples the attention and mlp weights--they no longer need to have the same shape / orientation.
3. It sets us up for cross-layer parameter sharing if we want to play with that. 

### Replicated Parameters

Something I realized is that, since we're communication bound, it's kind of abusrd to be dividing up the work on the smear gate (1, 12), attention gates (10, 6, 12), and scalars (62). 

There's a solution for that: we use a "replicated parameters" strategy. It just means that you issue an `all_reduce`, every GPU receives the full set of gradient updates for those three tiny parameters, and every GPU calculates the full weight update for them. Redundant work, but it means no `all_gather` step is required afterward. 

### Result

With all of the above changes, our step is looking something more like the below. 

<img width="2000" height="77" alt="image" src="https://github.com/user-attachments/assets/68daa552-41f9-407b-891b-eb905dcd0f5a" />

I'm still messing around with hook and compute ordering, and this is from some arbitrary trial, but note how there are no gaps in the comms! 🎉

Most of my runs have been foiled by not understanding how those backwards hooks really work--and I still kinda don't. For example, sometimes that first scatter takes a really long time, which must have something to do with alignment between the GPUs?

But one of my early trials shaved 300ms off my 12-11 record, so there's that much at least.

**Comment by chrisjmccormick** [comment]
## Combined Optimizer Benefits

One of the benefits of combining the optimizers is that we can make the handling of the different parameters very arbitrary, and have the logic for that in one place.

Right now, I have the configuration setup as a table, but we could move these properties to the parameters themselves to be more consistent with the current coding style.

```python
# Define a table of all labeled parameters and their desired settings.
# The list order dictates the order in which we will wait for the gradients to be ready.
# Optimizer: Use NorMuon for projections, Adam for scalars and embeddings.
#    Policy: 'replicated' means every GPU handles the full parameter, to avoid communication overhead.
#            'sharded' means the parameter is split into world_size chunks, and each GPU handles one chunk.
# Reduction Dim: Dictates the variance normalization by NorMuon. Choose -2 if the neurons are in rows, 
#               -1 if in columns, None for Adam / to skip it.
# View: Use this if your parameter needs to be "reshaped" before being presented to comms and to Muon. The leading dimension must be divisible by 8--pad it if necessary.
param_table = [
    # Label         Optimizer  Policy       Lr Mul  WD Mul  Red Dim  View            Sync%
    ("ve2",         "adam",    "sharded",   75.0,    1.0,    None,    None,               2),  # (50304, 768)
    ("lm_head",     "adam",    "sharded",    1.0,    1.0,    None,    None,               2),  # (50304, 768) - Slow to accumulate grads
    ("ve1",         "adam",    "sharded",   75.0,    1.0,    None,    None,               2),  # (50304, 768)
    ("ve0",         "adam",    "sharded",   75.0,    1.0,    None,    None,               2),  # (50304, 768)
    ("mlp",         "normuon", "sharded",    1.0,    1.0,    -1,      (24, 3072, 768),    1),  # (12, 2, 3072, 768) -> (24, 3072, 768)
    ("attn",        "normuon", "sharded",    1.0,    1.0,    -1,      (40, 768, 768),     1),  # (10, 3072, 768) -> (40, 768, 768)
    ("embed",       "adam",    "sharded",   75.0,    1.0,    None,    None,               2),  # (50304, 768)
    ("scalars",     "adam",    "replicated", 5.0,    1.0,    None,    None,               2),  # (62,)
    ("smear_gate",  "normuon", "replicated", 1.0,    1.0,    -1,      None,               1),  # (1, 12)
    ("attn_gate",   "normuon", "replicated", 1.0,    1.0,    -1,      None,               1),  # (10, 6, 12)
]
```

Most notably, the compute order can be arbitrary, and I also moved the "optimize every `x` steps" logic into the optimizer, so that we can mess around with how often to update each individual parameter.

I also made the reduction dim and the reshaping explicit. 

The reduction dim used to be a shape-based heurstic, and that tripped us up. 

**Specifying Shape**

The "view" setting replaces the "custom grouping" logic in the current code, and brings it to the front of our attention.

Something that I've found to be a bummer about the current codebase is that if you have a clever idea that requires changing the parameters of the model, the optimizer is going to fall back on a naive stacking strategy for your weights, and your run isn't going to win unless you know how to customize the batching algorithm to align with your idea.

That hasn't changed, but I think this makes it a little easier to solve.

Whether by padding or clever packing, you have to define your parameters in a way that can be reshaped such that the leading dimension is evenly divisible by 8. If you do that, then there's hope it will be optimized efficiently. 

And if you have to pad out your parameter a ton to meet the % 8 requirement, maybe rethink it a bit!

---

### Final Thoughts

I'm not sure if it's strictly necessary to move to the parameter bank approach over the current batching strategy. If the parameter bank is just too weird, we may be able to address the memcpy overhead just by combining the optimizers and re-ordering things for better overlap.

I'm putting a requirement on myself here that this all result in fewer lines of code and a clearer implementation. I'm optimistic that's the case, particularly if we're good with the per-parameter optimizer configuration being more explicit / brittle. The current implementation does some hand-holding--padding things for you, inferring behavior based on parameter shape, and sizing buffers only after it receives the actual workload. All of that adds lines and (I think) some complexity.

**Comment by shenberg** [comment]
small note regarding hooks, as @akash5474 noted in his blog post, torch compiled-autograd mode should be the solution to allow compiling the backwards hooks into the backwards graph so we can get the hooks firing as soon as gradients are ready and not after the compiled backward pass. He did say that it didn't work for him for some unknown reason. The other solution I saw when investigating this is the Compiled-DDP optimization, which is the older solution to this issue. There's an optimization path that adds graph breaks into the backwards pass in the correct positions in order to allow backward and comms overlap for models that use DDP (`torch._dynamo.config.optimize_ddp=True`). This seems somewhat unsupported based on how people are talking about it in github issues.

**Reply by YouJiacheng** [reply]
maybe @akash5474 used a torch version prior to this fix: https://github.com/pytorch/pytorch/pull/150074?
or some other bugs occur.

**Comment by YouJiacheng** [comment]
<img width="1320" height="550" alt="image" src="https://github.com/user-attachments/assets/7ec7b5d3-5127-4e3a-b290-c9513e6476f3" />

IIRC `register_post_accumulate_grad_hook` should work for `p.grad` from `None` to `Tensor`, not only real accumulation.

Compiled DDP uses this.

<img width="780" height="1230" alt="image" src="https://github.com/user-attachments/assets/31cc8bd7-8d01-443f-a6f6-952e7eed3c0d" />

**Comment by YouJiacheng** [comment]
see also: https://github.com/KellerJordan/modded-nanogpt/pull/97
